### PR TITLE
fix(sites-resolve): preserve no_entitlement_for_product for WAITING_FOR_IP_ALLOWLISTING when caller is internal

### PR DESCRIPTION
**Pull request by habansal**
## Summary
- Internal Adobe callers (org in `ASO_PLG_EXCLUDED_ORGS`) viewing a customer with `WAITING_FOR_IP_ALLOWLISTING` PLG onboarding status were seeing "No site onboarded" instead of the IP allowlist dialog
- Root cause: the `callerIsInternal` remap (`no_entitlement_for_product → site_not_enrolled`) was unconditional, firing even when a customer is mid-PLG-onboarding with a bot-blocked site
- Fix: when an internal caller resolves an org with **no customer-visible entitlement** — either none yet, or a `PRE_ONBOARD` one — check `PlgOnboarding.allByImsOrgId()` for any active `WAITING_FOR_IP_ALLOWLISTING` record; if found, preserve `no_entitlement_for_product` so `PlgAppGate` fires and shows the IP allowlist dialog
- Additional correctness fix: the PLG check is short-circuited for non-ASO products — `PlgOnboarding` records are ASO-only, so firing the DB query on LLMO/ACO requests was incorrect (could return the ASO PLG wizard error for orgs with no LLMO entitlement)

## Background
When a PLG customer hits the bot blocker (Step 4 of onboarding), the `PlgOnboarding` record is created with `WAITING_FOR_IP_ALLOWLISTING` status. The ASO entitlement isn't created until Step 10, so a mid-onboarding org has either **no entitlement** or a non-customer-visible **`PRE_ONBOARD`** entitlement (created during pre-onboarding). The existing `callerIsInternal` guard in `resolveSite` remapped `no_entitlement_for_product → site_not_enrolled` regardless, causing internal users to see "No site onboarded" for these customers.

Discovered during the datacom.com PLG onboarding incident (June 2026). Identified by Kanishka as the fix needed in `/sites-resolve`. Jira: [SITES-47189](https://jira.corp.adobe.com/browse/SITES-47189). Slack thread: https://cq-dev.slack.com/archives/C085NJRJG2W/p1782351621608179

## Changes
- `src/controllers/sites.js`:
  - added `isOrgWaitingForIpAllowlisting(org)` helper — short-circuits for non-ASO `productCode`, fail-open on DB error, null-safe on empty records
  - a single hoisted guard in **both** the `siteId` path and the `resolveByOrg` closure: for internal callers with no customer-visible entitlement (`!hasCustomerVisibleEntitlement` covers both no-entitlement and `PRE_ONBOARD`), preserve `no_entitlement_for_product` when the org is `WAITING_FOR_IP_ALLOWLISTING`
- `test/controllers/sites.test.js`: added `PlgOnboarding` mock to `mockDataAccess`; tests covering the bug scenario across no-entitlement and `PRE_ONBOARD` tiers, remap preservation, non-ASO product short-circuit, and fail-open/null-safe paths on both call sites

## Test plan
- [x] `internal caller, no entitlement + WAITING_FOR_IP_ALLOWLISTING (ASO) → 404 no_entitlement_for_product` (siteId path)
- [x] `internal caller, PRE_ONBOARD + WAITING_FOR_IP_ALLOWLISTING → 404 no_entitlement_for_product` (siteId path)
- [x] `internal caller, organizationId path: no entitlement + WAITING_FOR_IP_ALLOWLISTING → 404 no_entitlement_for_product`
- [x] `internal caller, organizationId path: PRE_ONBOARD + WAITING_FOR_IP_ALLOWLISTING → 404 no_entitlement_for_product`
- [x] `internal caller, no entitlement + non-WAITING records → 404 site_not_enrolled` (remap preserved)
- [x] `internal caller, non-ASO product + WAITING_FOR_IP_ALLOWLISTING → 404 site_not_enrolled` (PLG check skipped)
- [x] `internal caller, PlgOnboarding lookup throws → 404 site_not_enrolled` (fail-open, both paths)
- [x] `internal caller, PlgOnboarding returns null records → 404 site_not_enrolled` (null-safe, both paths)
- [x] All existing `resolveSite` tests pass unchanged
- [x] Customer callers completely unaffected
- [x] Branch-deploy manual validation (internal vs external caller, ASO vs LLMO product) — see comment below

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/controllers/sites.js [hunk 1]
```diff
@@ -1666,6 +1682,9 @@ function SitesController(ctx, log, env) {
 
       if (!entitlement) {
         if (callerIsInternal) {
+          if (await isOrgWaitingForIpAllowlisting(org)) {
```
Comment (MysticatBot) [inline src/controllers/sites.js]: issue (blocking): The `resolveByOrg` path silently returns `no_entitlement_for_product` without any log statement when `WAITING_FOR_IP_ALLOWLISTING` is detected, while the `siteId` path (line ~1739) logs at `info` level. This asymmetry makes debugging harder - when an internal caller hits this branch via the organizationId/imsOrg resolve path, there is no log entry explaining why the remap was suppressed.

Add a `log.info` matching the siteId-path message, substituting `orgId` or `imsOrgId` for `siteId`. Example:
```js
if (await isOrgWaitingForIpAllowlisting(org)) {
  log.info(`[resolveSite] Internal caller (callerImsOrg=${callerImsOrg}): WAITING_FOR_IP_ALLOWLISTING detected, preserving no_entitlement_for_product for org=${org.getId()}`);
  return resolveFailure(...);
}
```

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @habansal,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - one observability gap to close before merge.
**Complexity: MEDIUM** - API surface (controller path).
_Changes: Preserves `no_entitlement_for_product` resolve status for internal callers when the target org has a PLG onboarding record in `WAITING_FOR_IP_ALLOWLISTING` state (2 files)._

## Must fix before merge
1. [Important] Missing log statement in resolveByOrg path when WAITING_FOR_IP_ALLOWLISTING detected - `src/controllers/sites.js:1685` (details inline)

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Test cleanup - the trailing `mockDataAccess.PlgOnboarding.allByImsOrgId.resolves([])` inside two test bodies is redundant since `sandbox.restore()` in `afterEach` already resets all stubs. Harmless but dead code - `test/controllers/sites.test.js:6996`
- suggestion: The `isOrgWaitingForIpAllowlisting` DB call fires on every internal no-entitlement request. Consider short-circuiting with a product code check (if the resolved product is not ASO/PLG, skip the lookup) to avoid the extra query on non-PLG paths - `src/controllers/sites.js:1632`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 9m 35s | <b>Cost:</b> $4.88 | <b>Commit:</b> <code>d1a035b5031891b2c3f50358558dd2cf6e1ac12f</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @habansal,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - prior blocking finding resolved, no new blocking issues.
**Complexity: MEDIUM** - API surface (controller path).
_Changes: Preserves `no_entitlement_for_product` resolve status for internal callers when the target org has a PLG onboarding record in `WAITING_FOR_IP_ALLOWLISTING` state (2 files)._

## Previously flagged, now resolved
- Missing log statement in `resolveByOrg` path when `WAITING_FOR_IP_ALLOWLISTING` detected - now added in both call sites (commit 309c28b)

<details>
<summary>Non-blocking (1): minor issues and suggestions</summary>

- nit: Untested defensive branches in `isOrgWaitingForIpAllowlisting` - the `!hasText(imsOrgId)` early return and the `catch` branch have no test coverage (the 5 lines Codecov flagged). A single test like `PlgOnboarding.allByImsOrgId.rejects(new Error('DB timeout'))` verifying the fallback to `site_not_enrolled` would document the fail-open intent - `src/controllers/sites.js:1637`
- suggestion: The `isOrgWaitingForIpAllowlisting` DB call fires on every internal no-entitlement request regardless of product context. Consider short-circuiting with a product code check to avoid the extra query on non-PLG paths - `src/controllers/sites.js:1635`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 18s | <b>Cost:</b> $4.11 | <b>Commit:</b> <code>309c28b02a0701a8d051de9a9b2e5b7a3cc25328</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @habansal,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - three post-approval commits add edge-case test coverage and remove dead code; no new concerns.
**Complexity: MEDIUM** - API surface (controller path).
_Changes: Preserves `no_entitlement_for_product` resolve status for internal callers when the target org has a PLG onboarding record in `WAITING_FOR_IP_ALLOWLISTING` state (2 files)._

## Previously flagged, now resolved
- Untested defensive branches in `isOrgWaitingForIpAllowlisting` (`!hasText` early return and `catch` branch) - now covered by fail-open and null-safe tests (commits 45afb37, bb4d9c0)
- Unreachable `!PlgOnboarding` guard removed (commit 6802238) - correctly identified as dead code since `PlgOnboarding` is destructured at factory time

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- suggestion: The fail-open and null-records tests exercise only the siteId path. A symmetric test for the organizationId path (`getFirstEnrollment`) would close the coverage gap if the two call sites ever diverge - `test/controllers/sites.test.js:7196`
- suggestion: The `isOrgWaitingForIpAllowlisting` DB call fires on every internal no-entitlement request regardless of product context. Consider short-circuiting with a product code check to avoid the extra query on non-PLG paths - `src/controllers/sites.js:1635`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 23s | <b>Cost:</b> $3.88 | <b>Commit:</b> <code>6802238c423a496e121a9110ad8cb373eb64c4ed</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @habansal,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - prior suggestions addressed, no new blocking issues.
**Complexity: MEDIUM** - API surface (controller path).
_Changes: Adds ASO product-code short-circuit to skip PlgOnboarding lookup for non-ASO products, plus symmetric organizationId path tests for fail-open and null-safe scenarios (2 files)._

## Previously flagged, now resolved
- Product-code short-circuit to avoid unnecessary PlgOnboarding DB call on non-ASO requests - now implemented via `ASO_PRODUCT_CODE` guard (commit f6f4406)
- Symmetric organizationId path tests for fail-open and null-safe branches - now added (commit 492d519)

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Test "internal caller, no entitlement + other PlgOnboarding records (not WAITING)" does not set `context.pathInfo = { headers: { 'x-product': 'ASO' } }`, so `isOrgWaitingForIpAllowlisting` short-circuits at the product-code guard before reaching `records.some(...)`. The test passes correctly (outcome is `site_not_enrolled` either way) but exercises the product-code exit rather than the intended "records exist, none match WAITING" branch. One-line fix: add the ASO header to that test - `test/controllers/sites.test.js:7018`
- suggestion: Add `expect(mockDataAccess.PlgOnboarding.allByImsOrgId).to.not.have.been.called` to the "non-ASO product + WAITING" test to make the short-circuit assertion explicit and match the test title's claim ("PLG check skipped") - `test/controllers/sites.test.js:6994`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 21s | <b>Cost:</b> $4.95 | <b>Commit:</b> <code>492d519ce1f9e33aa794ac41a00ef5174036b761</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by ravverma — APPROVED** [review]
LGTM

**Comment by habansal** [comment]
## Branch Deploy Validation

Manually validated against the branch deploy at `https://spacecat.experiencecloud.live/api/ci/` using a mysticat token (AEM Sites Engineering org — confirmed internal, i.e. in `ASO_PLG_EXCLUDED_ORGS` on CI).

| # | Scenario | callerImsOrg | x-product | Expected | Actual |
|---|----------|--------------|-----------|----------|--------|
| 1 | ASO PAID org — happy path | — | ASO | 200 with site data | ✅ 200 |
| 2 | ASO FREE_TRIAL org — internal caller, skips tier check | AEM Sites Eng | ASO | 200 with site data | ✅ 200 |
| 3a | LLMO-only org, external caller | — | ASO | 404 `no_entitlement_for_product` | ✅ `no_entitlement_for_product` |
| 3b | LLMO-only org, internal caller | AEM Sites Eng | ASO | 404 `site_not_enrolled` | ✅ `site_not_enrolled` |
| 4 | LLMO-only org, internal caller | AEM Sites Eng | LLMO | 404 `site_not_enrolled` (no enrolled site) | ✅ `site_not_enrolled` |
| 5 | ASO PAID org, internal caller | AEM Sites Eng | LLMO | 200 with site data | ✅ 200 |

**Note:** The specific `WAITING_FOR_IP_ALLOWLISTING` fix (internal caller should see `no_entitlement_for_product` instead of `site_not_enrolled` when the org is WAITING) could not be exercised on the CI database — the only WAITING org on dev already has an ASO PAID entitlement and never reaches the no-entitlement path. That path is covered by the 8 unit tests in the test plan above.

All observable code paths validate correctly. No regressions detected.

File: src/controllers/sites.js [hunk 2]
```diff
@@ -1666,6 +1681,10 @@ function SitesController(ctx, log, env) {
 
       if (!entitlement) {
```
Comment (Kanishkavijay39) [inline src/controllers/sites.js:1701]: When a site is pre-onboarded, it has a PRE_ONBOARD entitlement. So !entitlement is false, and the entire isOrgWaitingForIpAllowlisting check is skipped. The code falls through to the PRE_ONBOARD tier path and returns site_not_enrolled instead of no_entitlement_for_product.

I think we should also add this check in the other if block. Maybe we can move this out of both if/else blocks:
if (!CUSTOMER_VISIBLE_TIERS.includes(entitlement.getTier())) {

**Review by Kanishkavijay39 — CHANGES_REQUESTED** [review]

**Review by MysticatBot — APPROVED** [review]
Hey @habansal,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - prior reviewer feedback addressed, no new blocking issues.
**Complexity: HIGH** - medium diff; API surface (controller path).
_Changes: Hoists the WAITING_FOR_IP_ALLOWLISTING guard into a combined `hasCustomerVisibleEntitlement` predicate covering both no-entitlement and PRE_ONBOARD tiers, addressing Kanishkavijay39's review feedback (2 files)._

## Previously flagged, now resolved
- PRE_ONBOARD tier bypassing the WAITING check (Kanishkavijay39's review) - now handled via `hasCustomerVisibleEntitlement` guard in both paths (commits ad3a1f5, dbc04d5)
- Product-code short-circuit to avoid unnecessary PlgOnboarding DB call on non-ASO requests - implemented in earlier commit
- Symmetric organizationId path tests for fail-open and null-safe branches - added in earlier commit

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Test "internal caller, no entitlement + other PlgOnboarding records (not WAITING)" does not set `context.pathInfo = { headers: { 'x-product': 'ASO' } }`, so `isOrgWaitingForIpAllowlisting` short-circuits at the product-code guard before reaching `records.some(...)`. The test passes correctly but exercises the product-code exit, not the "records exist, none match WAITING" branch. One-line fix: add the ASO header to that test - `test/controllers/sites.test.js:7018`
- suggestion: Add a symmetric "other records (not WAITING)" test for the organizationId path with `x-product: ASO` set, to guard against future divergence between the two resolve paths - `test/controllers/sites.test.js:7277`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 39s | <b>Cost:</b> $4.55 | <b>Commit:</b> <code>dbc04d5162032fc1b3c1e6f9bb7bc9c7bd169a6f</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/controllers/sites.js (hunk 2 above)
Comment (habansal) [reply src/controllers/sites.js:1701]: @Kanishkavijay39 Thanks for catching this.
I have addressed this issue. Can you re-review?

File: src/controllers/sites.js [hunk 3]
```diff
@@ -1672,6 +1687,17 @@ function SitesController(ctx, log, env) {
       const tierClient = TierClient.createForOrg(context, org, productCode);
       const { entitlement, site: enrolledSite } = await tierClient.getFirstEnrollment();
 
+      // Internal caller viewing a customer with no customer-visible entitlement
+      // (none yet, or PRE_ONBOARD) who is mid-PLG-onboarding (WAITING_FOR_IP_ALLOWLISTING):
+      // preserve no_entitlement_for_product so PlgAppGate shows the IP allowlist dialog.
+      const hasCustomerVisibleEntitlement = entitlement
```
Comment (Kanishkavijay39) [inline src/controllers/sites.js:1693]: This check is not needed how far i can see, as isOrgWaitingForIpAllowlisting will handle so instead of 3 and we can make it simple with two && that's 
(callerIsInternal &&  await isOrgWaitingForIpAllowlisting(org))

**Review by Kanishkavijay39 — APPROVED** [review]

File: src/controllers/sites.js (hunk 3 above)
Comment (habansal) [reply src/controllers/sites.js:1693]: This can happen in theory: `sites.js`/`resolveSite` is independent of the plg-onboarding lifecycle, so it can't assume a WAITING org has no entitlement. `isOrgWaitingForIpAllowlisting` is org-scoped (any WAITING record for the imsOrg), and an org can hold a customer-visible entitlement from one domain while a different domain sits in WAITING — e.g. on dev, org `44568c3e` has ASO PAID *and* a WAITING record for `labcorp.com` at the same time. Without `!hasCustomerVisibleEntitlement`, an internal caller resolving such an org would get `no_entitlement_for_product` despite a valid entitlement, so the check is needed.
